### DISCUSSION TITLE
Multi-threading behavior with v7 client

### DISCUSSION THREAD

**Discussion by qrli** [rabbitmq/rabbitmq-dotnet-client#1663] [Q&A]
In previous versions,  the doc states that a model/channel must be only accessed from one thread at a time. 

In the v7 implementation, I see that the .NET Channel classes are used internally and there is an internal thread to actually process and send commands to server. My gut-feeling is that now it is safe for my code to call methods on a RabbitMQ model/channel object from different threads without locks in my code. On the other hand, the async API makes it impractical to do locks, where the lock will span over awaits which is a recipe for deadlock.

Just want to confirm whether my understanding is correct. Thanks in advance.

**Comment by lukebakken** [comment]
> On the other hand, the async API makes it impractical to do locks, where the lock will span over awaits which is a recipe for deadlock.

Use `SemaphoreSlim` instead of `lock`. This code base uses `SemaphoreSlim` in several places.

> My gut-feeling is that now it is safe for my code to call methods on a RabbitMQ model/channel object from different threads without locks in my code.

What would be extremely useful to this project is for you to use the latest RC (https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-rc.8), remove the locks from your code, and report back.

After a quick review of the `SocketFrameHandler` class, I think you're correct. I will add some tests to confirm.

**Reply by qrli** [reply]
> Use SemaphoreSlim instead of lock. This code base uses SemaphoreSlim in several places.

Ahh, that's great to know. I'll have a try.

**Comment by michaelklishin** [comment]
For multi-frame operations (namely `basic.publish`) and synchronous (as in, having a protocol response) protocol methods (e.g. `queue.declare` or `queue.bind`) it is a extremely likely that even with socket-level synchronization in the library you will run into obscure issues, async API or not.

Bunny (the most popular Ruby client) does exactly this: it synchronizes socket writes because not doing so is not practical in the context of certain Web servers, but sharing channels between threads is not supported. Two levels of locking on the hot code path and the complexity of a concurrent continuation implementation (in Ruby, it can be much easier to reason about in C#) were not worth the effort. My point is that socket-level synchronization is not enough in practice with key operations mentioned above.

**Reply by lukebakken** [reply]
Good point. My guess is that `basic.publish` would be OK, but I have no idea what would happen if a shared `IChannel` instance registers consumers across threads, or what might happen during connection/topology recovery.

**Reply by qrli** [reply]
Good to know. Thanks.

I don't do consuming side in multi-threads and no reason to do so. My case is for message senders - multiple threads (from HTTP requests) are producing messages. When I was trying to serialize that into a single thread with .NET's Channel, I found the RabbitMQ client library is already using Channel internally, thus the question. I'll take your advice and try SemaphoreSlim.

**Reply by lukebakken** [reply]
You should benchmark what happens if you create a new `IChannel` instance per HTTP request. You should also consider using an `ObjectPool` to cache existing `IChannel` instances.

Finally, I'm 99.99% sure that publishing is thread-safe when using a shared `IChannel` instance.

**Reply by rhuijben** [reply]
Haven't seen issues in our application, using a single channel now. 
(We switched from EasyNetQ's advanced queue code to directly using the v7 api recently.)

**Reply by lukebakken** [reply]
@rhuijben thank you for chiming in!

What are you doing with the shared channels? Just publishing? Consuming?

**Reply by rhuijben** [reply]
Publishing and consuming.

I just use a single channel on a single connection per service instance now, and this has been more stable+a bit faster than our previous implementation that had many models over two connections, for the last few weeks.

The performance is probably mostly removing unneeded layering though.

**Reply by dario-l** [reply]
@rhuijben Exactly the same refactoring in our code base we want to do. For now we use EasyNetQ but when we try to run partitioning (with consistent hash exchange) we end up with a lot of nasty bugs.

I was wandering is really one channel enough for every operation? I found this: https://www.rabbitmq.com/docs/channels#flow-control
That's why we are want to use min 2 connections - one for publishing and one for consuming.

We have modular monolith application with well defined and fine grained modules (almost 60 for now).
We are planning to create two connections:
- one for publishers - where publishers from different modules could rent a channel from a pool
- one for consumers - where one queue will have one channel for all subscribers from that queue

Is that sounds reasonable or is it a overkill? 😇

**Reply by lukebakken** [reply]
@dario-l yes, your publishers and consumers should be on two separate _connections_, like you state.

**Reply by dario-l** [reply]
@lukebakken thanks for confirming my decision. 😃

**Reply by michaelklishin** [reply]
@dario-l a few relevant doc sections, both to help you understand the proposed idea of using separate connections for publishers and consumers, and observe publisher blocking events that usually drive the decision to separate the two:

1. https://www.rabbitmq.com/docs/connections#flow-control
2. https://www.rabbitmq.com/docs/connection-blocked
3. https://www.rabbitmq.com/docs/connections#client-provided-names

**Reply by danielmarbach** [reply]
Besides having a dedicated connection for publishers and consumers, have we ever settled on the guidance around sharing the publisher channel across multiple concurrent operations? According to my knowledge and also mentioned again by @lukebakken this should be fine since the publish operations are being added to the socket frame handler using a System.Threading.Channel and then dispatched. 

That would indicate as long as you have consuming, admin and publishing separated there should be no interference happening and there is no reason to have multiple pooled channels anymore for the "concurrent" publishes. Should the guidance be refined?

**Reply by lukebakken** [reply]
When someone ask I tell them that the `IChannel` instances are thread-safe. I'm assuming that you refer to [this web page](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#concurrency) when it comes to "official guidance"?

**Reply by SaifAqqad** [reply]
> That would indicate as long as you have consuming, admin and publishing separated there should be no interference happening and there is no reason to have multiple pooled channels anymore for the "concurrent" publishes. Should the guidance be refined?

This should be mentioned clearly in the [dotnet client docs](https://www.rabbitmq.com/client-libraries/dotnet-api-guide#publishing), as the current state of the client docs + rabbitmq server/generic docs have very conflicting guidance

I would also suggest adding more examples that implement this usage in a more realistic scenario (for example in an ASP.NET Core web API)

**Reply by lukebakken** [reply]
Please take the time to open pull requests to do the work you are suggesting. If you don't have time to do the work yourself, at least point out _exactly_ where the issues are, rather than make statements such as this:

>  as the current state of the client docs + rabbitmq server/generic docs have very conflicting guidance

> I would also suggest adding more examples that implement this usage in a more realistic scenario (for example in an ASP.NET Core web API)

Please feel free to do this yourself. The tutorial code can be found here:

https://github.com/rabbitmq/rabbitmq-tutorials/tree/main/dotnet

...and the associated website tutorials can be found here:

https://github.com/rabbitmq/rabbitmq-website/tree/main/client-libraries

https://github.com/rabbitmq/rabbitmq-website/tree/main/tutorials

My time to focus on the .NET client is _extremely_ limited these days, so anything that the community can do on its own is appreciated.

**Reply by SaifAqqad** [reply]
> If you don't have time to do the work yourself, at least point out exactly where the issues are

This is from the dotnet client docs:
> This is a hard requirement for publishers: sharing a channel (an IChannel instance) for concurrent publishing will lead to incorrect frame interleaving at the protocol level. Channel instances must not be shared by threads that publish on them.

However, based on the comments made in this discussion, it seems v7 does not require separate (or pooled) channels for concurrent publishes, so the docs should probably be updated.
________________
> rather than make statements such as this

I'm sorry, my comment was not directed at the maintainers or anyone specific, it was just a general suggestion.
